diff --git a/Makefile b/Makefile
index 3729601..c74e5aa 100644
--- a/Makefile
+++ b/Makefile
@@ -119,14 +119,19 @@ SUBDIRS+=lua-signal
 endif
 
 CHDKPTP_EXE=chdkptp$(EXE_EXTRA)$(EXE)
+CHDKPTP_LIB=chdkptp$(EXE_EXTRA)$(LIB)
 
 EXES=$(CHDKPTP_EXE)
+LIBS=$(CHDKPTP_LIB)
 
-all: $(EXES)
+all: $(EXES) $(LIBS)
 
 SRCS=properties.c ptp.c chdkptp.c lbuf.c liveimg.c rawimg.c luautil.c $(PTPIP_SRCS)
 OBJS=$(SRCS:.c=.o)
 
+$(CHDKPTP_LIB): $(OBJS)
+	$(CC) -shared -o $@ lfs/lfs.o $^ $(LDFLAGS)
+
 $(CHDKPTP_EXE): $(OBJS)
 	$(CC) -o $@ lfs/lfs.o $^ $(LDFLAGS)
 
diff --git a/chdkptp.c b/chdkptp.c
index 47cbba2..bc99211 100644
--- a/chdkptp.c
+++ b/chdkptp.c
@@ -2193,7 +2193,11 @@ static const luaL_Reg lua_errlib[] = {
 };
 
 
-static int chdkptp_registerlibs(lua_State *L) {
+int luaopen_chdkptp(lua_State *L) {
+	usb_init();
+	luaopen_lfs(L);
+	luaopen_lbuf(L);
+	luaopen_rawimg(L);
 	/* set up meta table for error object */
 	luaL_newmetatable(L,CHDK_API_ERROR_META);
 	lua_pushcfunction(L,api_error_tostring);
@@ -2262,13 +2266,9 @@ int main(int argc, char ** argv)
 	g_argv = argv;
 	/* register signal handlers */
 //	signal(SIGINT, ptpcam_siginthandler);
-	usb_init();
 	lua_State *L = luaL_newstate();
 	luaL_openlibs(L);
-	luaopen_lfs(L);
-	luaopen_lbuf(L);
-	luaopen_rawimg(L);	
-	chdkptp_registerlibs(L);
+	luaopen_chdkptp(L);
 	int r=exec_lua_string(L,"require('main')");
 	uninit_gui_libs(L);
 	lua_close(L);
diff --git a/include.mk b/include.mk
index 2b9a409..da0fb5e 100644
--- a/include.mk
+++ b/include.mk
@@ -8,6 +8,8 @@ ifeq ($(HOSTPLATFORM),MINGW)
 OSTYPE=Windows
 EXE=.exe
 DLL=.dll
+LIB=.dll
+
 # Note may be freetype or freetype6 depending on your CD version, zlib requried for 5.5 and later
 CD_FREETYPE_LIB=freetype6 z
 #CD_FREETYPE_LIB=freetype z
@@ -16,6 +18,8 @@ ifeq ($(HOSTPLATFORM),Linux)
 OSTYPE=Linux
 EXE= 
 DLL=.so
+LIB=.so
+
 CD_FREETYPE_LIB=freetype z
 endif
 endif
@@ -24,7 +28,7 @@ endif
 EXE_EXTRA=
 
 CC=gcc
-CFLAGS=-DCHDKPTP_OSTYPE=\"$(OSTYPE)\" -Wall
+CFLAGS=-DCHDKPTP_OSTYPE=\"$(OSTYPE)\" -Wall -fPIC -g
 LDFLAGS=
 #LD=ld
 
